### PR TITLE
Add TypeScript definition and require Node.js 6

### PR DESCRIPTION
**Pull request by DenisCarriere**
Add TypeScript definition to `write-json-file`

Ref: https://github.com/sindresorhus/typescript-definition-style-guide
https://github.com/sindresorhus/ama/issues/439#issuecomment-414212229

### CODE REVIEW

**Comment by sindresorhus** [comment]
Thanks for creating a type definition 🙌

**Comment by DenisCarriere** [comment]
@sindresorhus You're welcome 😄 I've been looking forward to add Typescript definitions to your repos for a while.

Glad you've started accepting them 👍

**Comment by DenisCarriere** [comment]
Still getting a few errors when running tests, not too sure if these are valid errors. 🤔 

1. `xo` doesn't seem to handle Typescript definitions very well
2. `prefer-destructuring`  doesn't seem like it's possible to deconstruct on line 52

CC: @sindresorhus 

```
$ npm t

> write-json-file@2.3.0 test /Users/denis/Github/write-json-file
> xo && ava


  index.d.ts:1:6
  ✖   1:6  Parsing error: Unexpected token Replacer

  index.js:52:4
  ✖  52:4  Use object destructuring.                 prefer-destructuring

  2 errors
npm ERR! Test failed.  See above for more details.
```

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+export function sync(filepath: string, data: any, options?: Options): void;
+
+/**
+ * Stringify and write JSON to a file atomically
+ *
+ * Creates directories for you as needed.
+ *
+ * @param {string} filepath Filepath
+ * @param {any} data Data
+ * @param {object} [options] Optional parameters
```
Comment (sindresorhus) [inline index.d.ts]: No need to have a description for these 3 parameters as the name is clear enough.

**Comment by sindresorhus** [comment]
> xo doesn't seem to handle Typescript definitions very well

Yeah... You have to exclude the type definition file from being linted because of [`xojs/xo#348`](https://github.com/xojs/xo/issues/348). [Example.](https://github.com/sindresorhus/delay/blob/c17e29b13c20e6be35d64d75f45ce174bb793158/package.json#L48-L52)

File: index.js [hunk 2]
```diff
@@ -67,6 +67,12 @@ module.exports = (fp, data, opts) => {
 		.then(() => init(main, fp, data, opts));
 };
 
+// Support for Typescript default export
+module.exports.default = (fp, data, opts) => {
+	return makeDir(path.dirname(fp), {fs})
+		.then(() => init(main, fp, data, opts));
+};
```
[the remark this answers is not in the document]
Comment (DenisCarriere) [reply index.js]: 👍 makes sense

File: package.json [hunk 3]
```diff
@@ -16,8 +16,10 @@
     "test": "xo && ava"
   },
   "files": [
-    "index.js"
+    "index.js",
+    "index.d.ts"
   ],
+  "types": "index.d.ts",
```
[the remark this answers is not in the document]
Comment (DenisCarriere) [reply package.json]: Sounds good, didn't know this wasn't required, good to know.

**Comment by sindresorhus** [comment]
I'm also looking for feedback on the style guide. Anything that could be clearer/improved? Anything you strongly disagree with? You should probably read it again as I have updated it after reviewing this PR.

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+ * @param {boolean|function} [options.sortKeys=false] Sort the keys recursively. Optionally pass in a compare function.
+ * @param {function|Array<number|string>|null} [options.replacer] Passed into JSON.stringify.
+ * @param {number} [options.mode=0o666] Mode used when writing the file.
+ * @returns {Promise<void>}
+ * @example
+ * const writeJsonFile = require('write-json-file');
+ *
+ * writeJsonFile('foo.json', {foo: true}).then(() => {
+ * 	console.log('done');
+ * });
```
[the remark this answers is not in the document]
Comment (DenisCarriere) [reply index.d.ts]: 👍 I like this example as well

File: index.d.ts (hunk 1 above)
Comment (DenisCarriere) [reply index.d.ts]: Fair enough, dropping those params can be extracted from the Types

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+ * Creates directories for you as needed.
+ *
+ * @param {string} filepath Filepath
+ * @param {any} data Data
+ * @param {object} [options] Optional parameters
+ * @param {string|number|null} [options.indent='\t'] Indentation as a string or number of spaces. Pass in null for no formatting.
+ * @param {boolean} [options.detectIndent=false] Detect indentation automatically if the file exists.
+ * @param {boolean|function} [options.sortKeys=false] Sort the keys recursively. Optionally pass in a compare function.
+ * @param {function|Array<number|string>|null} [options.replacer] Passed into JSON.stringify.
+ * @param {number} [options.mode=0o666] Mode used when writing the file.
```
[the remark this answers is not in the document]
Comment (DenisCarriere) [reply index.d.ts]: 👍

File: package.json [hunk 6]
```diff
@@ -10,13 +10,14 @@
     "url": "sindresorhus.com"
   },
   "engines": {
-    "node": ">=4"
+    "node": ">=6"
   },
   "scripts": {
-    "test": "xo && ava"
+    "test": "xo --ignore index.d.ts && ava"
```
Comment (sindresorhus) [inline package.json]: Do it like the example is shared: https://github.com/sindresorhus/delay/blob/c17e29b13c20e6be35d64d75f45ce174bb793158/package.json#L48-L52

File: index.d.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+ * @example
+ * import * as writeJsonFile from 'write-json-file';
+ *
+ * writeJsonFile.sync('foo.json', {foo: true});
+ * console.log('done');
+ */
+export function sync(filepath: string, data: any, options?: Options): void;
+
+/**
+ * Stringify and write JSON to a file atomically
```
Comment (sindresorhus) [inline index.d.ts]: Should end in a dot.

File: index.d.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * Creates directories for you as needed.
+ *
+ * @returns {Promise<void>}
```
Comment (sindresorhus) [inline index.d.ts]: This is moot as it has no description and therefore is better documented by the type definition itself.

File: package.json (hunk 6 above)
Comment (DenisCarriere) [reply package.json]: 👍 Sounds good!

File: index.d.ts [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+	 */
+	mode?: number;
+}
+
+/**
+ * Stringify and write JSON to a file atomically
+ *
+ * Creates directories for you as needed.
+ *
+ * @returns {void}
```
Comment (sindresorhus) [inline index.d.ts]: Same as https://github.com/sindresorhus/write-json-file/pull/17/files#r211397928

File: index.d.ts [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+ * @example
+ * import * as writeJsonFile from 'write-json-file';
```
Comment (sindresorhus) [inline index.d.ts]: Should be `import writeJsonFile from 'write-json-file';`

File: index.d.ts (hunk 8 above)
Comment (DenisCarriere) [reply index.d.ts]: 👍 yea it's pretty pointless, it is defined by the Types

File: index.d.ts [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+ * Stringify and write JSON to a file atomically.
+ *
+ * Creates directories for you as needed.
+ *
+ * @returns {Promise<void>}
+ * @example
+ * import writeJsonFile from 'write-json-file';
+ *
+ * (async () => {
+ * 		await writeJsonFile('foo.json', {foo: true});
```
Comment (sindresorhus) [inline index.d.ts]: Incorrect indent

**Comment by sindresorhus** [comment]
Need feedback on: https://github.com/sindresorhus/write-json-file/pull/17#discussion_r211386978

> Is there any way to define default arguments in the types themselves instead of in the docs?

We need to define default values somewhere for improved auto-completion experience.

File: index.d.ts (hunk 10 above)
Comment (DenisCarriere) [reply index.d.ts]: 🤔 ... to access `.sync()` you need to add `import * as`

File: index.d.ts (hunk 11 above)
Comment (DenisCarriere) [reply index.d.ts]: 🤦‍♂️

**Comment by sindresorhus** [comment]
// @brandon93s @NeekSandhu Would you be able to help review this?

File: index.d.ts (hunk 10 above)
Comment (sindresorhus) [reply index.d.ts]: Ah, good point 👍

File: index.d.ts [hunk 12]
```diff
@@ -0,0 +1,54 @@
+type Replacer = (key: string, value: any) => void;
+
+interface Options {
+	indent?: string | number | null;
+	detectIndent?: boolean;
```
[the remark this answers is not in the document]
Comment (DenisCarriere) [reply index.d.ts]: Good question, I don't think it's possible as an interface.

```
[ts] A parameter initializer is only allowed in a function or constructor implementation.
```

You can do it if the types are created from a `.ts` file

File: index.d.ts (hunk 10 above)
Comment (DenisCarriere) [reply index.d.ts]: I guess if you use Babel this example is correct, I'll change it to:

```js
import writeJsonFile from 'write-json-file';
```

**Comment by DenisCarriere** [comment]
@sindresorhus I've looked over all your comments and I hope we've covered everything 😄 

Got another PR https://github.com/sindresorhus/load-json-file/pull/15 with the same changes.

I'm done for commits, unless you have more comments and reviews.

Let me know if you come up with a solution for https://github.com/sindresorhus/write-json-file/pull/17#discussion_r211386978

File: index.d.ts (hunk 10 above)
Comment (DenisCarriere) [reply index.d.ts]: Most Typescript users kinda got used to adding `import * as`, no need to include it.

Most docs are mixed between Babel & Typescript, we can keep as the Babel way

**Comment by brandon93s** [comment]
Looks good to me. 

For https://github.com/sindresorhus/write-json-file/pull/17#discussion_r211386978, unfortunately I don't think there is a good way for defaults to be defined on an interface. Adding the default to the interface documentation would at least make it visible in the intellisense: 

```js
/**
 * Detect indentation automatically if the file exists.
 * 
 * Default: false
 */
detectIndent?: boolean;
```

image

File: index.d.ts [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ * Creates directories for you as needed.
+ *
+ * @example
+ * import writeJsonFile from 'write-json-file';
+ *
+ * writeJsonFile.sync('foo.json', {foo: true});
+ * console.log('done');
+ */
+export function sync(filepath: string, data: any, options?: Options): void;
```
Comment (brandon93s) [inline index.d.ts]: `data` can't _really_ be `any`. For example, passing in `undefined` will throw a `TypeError`. We could attempt to express valid JSON values to be more specific here.

Comment (NeekSandhu) [reply index.d.ts]: Consider the fact even in TypeScripts "official" `lib.d.ts` file, the signature for `JSON.stringify( ... )` is

```typescript
JSON.stringify(value: any, replacer?: (key: string, value: any) => any, ...): string;
```
That means even the TypeScript team couldn't come with a reliabl, sound way to describe JSON.

File: package.json [hunk 14]
```diff
@@ -44,5 +45,10 @@
     "ava": "*",
     "tempfile": "^2.0.0",
     "xo": "*"
-  }
+  },
+  "xo": {
+		"ignores": [
```
Comment (SamVerschueren) [inline package.json]: The indentation is off. These lines are using tabs and the others are using spaces.

I know @sindresorhus switched to tabs for package.json because npm preserves them nowadays. But think it should be done in a separate commit.

**Comment by NeekSandhu** [comment]
As far as https://github.com/sindresorhus/write-json-file/pull/17#discussion_r211386978 goes, like @brandon93s said, unfortunately not possible yet. I've been using the method he advised, i.e to describe the default value in comment(s), works really well IMO.

File: package.json (hunk 14 above)
Comment (DenisCarriere) [reply package.json]: 👍

**Comment by DenisCarriere** [comment]
Added defaults using the `Default: '\t'` schema in the comments.

File: index.d.ts (hunk 13 above)
Comment (DenisCarriere) [reply index.d.ts]: There's always the `object` type, however I kinda like using `any` for this param.

Comment (NeekSandhu) [reply index.d.ts]: @brandon93s was hoping to prevent passing `undefined` (which causes `TypeError`), the problem is that no signature of any kind (as of `ts@3.0.1`) can prevent that from happening.

```typescript
function foo(a: object | number | string): void { }

// ts won't complain i.e no red squiggly lines
foo(undefined)
// no complains here either
foo(null)
```

**Comment by sindresorhus** [comment]
According to https://twitter.com/felixfbecker/status/1031660800413454337, we can use the `@default` tag, so let's do that.

File: index.d.ts (hunk 13 above)
Comment (sindresorhus) [reply index.d.ts]: Yeah, let's make it strict. Can be a type called `JSONStringifyable` maybe.

**Comment by NeekSandhu** [comment]
wow...totally TIL thing. It's even worth mentioning in your style guide.

File: .gitignore [hunk 15]
```diff
@@ -1 +1,2 @@
 node_modules
+yarn.lock
```
Comment (SamVerschueren) [inline .gitignore]: Add newline at the end of the file.

**Comment by NeekSandhu** [comment]
While the context is heated up, have you considered https://github.com/sindresorhus/ama/issues/554

File: .gitignore (hunk 15 above)
Comment (DenisCarriere) [reply .gitignore]: Thanks for catching this 👍

**Comment by DenisCarriere** [comment]
@SamVerschueren @sindresorhus 
- [x] Added new line to `.gitignore`
- [x] Using `@default` tags (ref: http://usejsdoc.org/tags-default.html)

File: index.d.ts (hunk 13 above)
Comment (DenisCarriere) [reply index.d.ts]: 👍 for making it more strict in the `data` param.

There's two options:

1. We create a type called `JSONStringifyable` (the type is slightly "hidden" when using intellisense)

screen shot 2018-08-21 at 9 57 48 am

2. We add the types directly in the `data` param, that way it's clear that the input types are `string`|`object`|`number` (my preferred choice).

screen shot 2018-08-21 at 9 59 41 am

File: index.d.ts [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * Creates directories for you as needed.
+ *
+ * @example
+ * import * as writeJsonFile from 'write-json-file';
+ *
+ * writeJsonFile.sync('foo.json', {foo: true});
+ * console.log('done');
+ */
+export function sync(filepath: string, data: object | number | string, options?: Options): void;
```
Comment (sindresorhus) [inline index.d.ts]: What do you think about making `object | number | string` a named type called `JSONStringifyable`? I think that would make it slightly more readable as it would explain why it randomly only accepts object/number/string.

File: index.d.ts [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+	 *
+	 * @default false
+	 */
+	detectIndent?: boolean;
+	/**
+	 * Sort the keys recursively. Optionally pass in a compare function.
+	 *
+	 * @default false
+	 */
+	sortKeys?: SortKeys | boolean;
```
Comment (sindresorhus) [inline index.d.ts]: I think `boolean` should be first as it should be sorted by most common usage, and most users will only ever need to set `{sortKeys: true}`.

Comment (DenisCarriere) [reply index.d.ts]: 👍 sounds good to me

File: index.d.ts (hunk 16 above)
Comment (DenisCarriere) [reply index.d.ts]: Ok! I'll add it

File: .gitignore (hunk 15 above)
Comment (DenisCarriere) [reply .gitignore]: Added newline

**Comment by sindresorhus** [comment]
This is perfect now. Thanks for your great work and patience, @DenisCarriere 🙌
